src/modules/esl-utils/decorators/test/prop.test.ts

```diff
@@ -1,22 +1,108 @@
+
+    const el = new TestClass();
+    expect(el.field).toBe('123');
+    expect(Object.getOwnPropertyDescriptor(el, 'field')).not.toBeNull();
+  });
+
+  describe('Provider based property resolves value trough it', () => {
+    const provider = jest.fn();
+    class TestClass {
+      @prop(provider)
+      public field: any;
+
+      @prop(provider, {readonly: true})
+      public readonlyField: any;
+
+      @prop(provider, {enumerable: true})
+      public enumerableField: any;
+    }
+
+    test('If provider function passed - value resolves trough it', () => {
+      const obj = new TestClass();
+      provider.mockReturnValue('initial');
+      expect(obj.field).toBe('initial');
+      provider.mockReturnValue('new');
+      expect(obj.field).toBe('new');
+    });
+
+    test('If provider function passed - provider receives instance as a context and argument', () => {
+      provider.mockImplementation(function() { return this; });
+      const obj = new TestClass();
+      expect(obj.field).toBe(obj);
+      expect(provider).toHaveBeenCalledWith(obj);
+    });
+
+    test('If provider function passed - resolver declared on prototype level', () => {
+      const obj = new TestClass();
+      expect(Object.hasOwnProperty.call(obj, 'field')).toBe(false);
+      expect(Object.getOwnPropertyDescriptor(TestClass.prototype, 'field')).not.toBeNull();
+    });
+
+    test('If provider function passed - value can be overwritten on instance', () => {
+      const obj = new TestClass();
+      const obj2 = new TestClass();
+      provider.mockReturnValue('initial');
+      obj.field = 'new';
+      expect(obj.field).toBe('new');
+      expect(obj2.field).toBe('initial');
+    });
+
+    test('If provider function passed - overwritten value could be cleared with delete', () => {
+      const obj = new TestClass();
+      provider.mockReturnValue('initial');
+      obj.field = 'new';
+      expect(obj.field).toBe('new');
+      delete obj.field;
+      expect(obj.field).toBe('initial');
+    });
+
+    test('If provider function passed - readonly property should ignores overwritten value', () => {
+      const obj = new TestClass();
+      provider.mockReturnValue('initial');
+      obj.readonlyField = 'new';
+      expect(obj.readonlyField).toBe('initial');
+    });
+
+    test('If provider function passed - property is not enumerable by default', () => {
+      expect(Object.keys(TestClass.prototype)).not.toContain('field');
+    });
+
+    test('If provider function passed - enumerable property should be enumerable', () => {
+      expect(Object.keys(TestClass.prototype)).toContain('enumerableField');
+    });
+  });
+
+  test('Overriding own property produce error', () => {
+    expect(() => {
+      class TestElement extends HTMLElement {
+        @prop({value: ''})
+        @attr()
+        public field: string;
+      }
+      new TestElement();
+    }).toThrowError(/own property/);
+  });
+
+  describe('Overriding @attr works fine', () => {
```

Comment:
This looks like an assertion, not a describe of the tests.
MB
```suggestion
  describe('Checking the @attr overriding', () => {
```

Reply:
BTW: ideal test should have a Test Case name, so it is statement but not the status or action
So I'd say both of our versions are not descriptive enough. So let's keep that update for now.

Reply:
ok